src/beacon/drand.rs

```diff
@@ -192,103 +195,115 @@ pub struct BeaconEntryJson {
-        if prev.round() == 0 {
-            return Ok(true);
-        }
+    fn verify_entries<'a>(
+        &self,
+        entries: &'a [BeaconEntry],
+        mut prev: &'a BeaconEntry,
+    ) -> Result<bool, anyhow::Error> {
+        let mut validated = vec![];
+        let is_valid = if self.network.is_unchained() {
+            let mut messages = vec![];
+            let mut signatures = vec![];
+            let pk = PublicKeyOnG2::from_bytes(&self.public_key)?;
+            {
+                let cache = self.verified_beacons.read();
+                for entry in entries.iter() {
+                    if cache.contains(&entry.round()) {
+                        continue;
+                    }
+
+                    messages.push(BeaconEntry::message_unchained(entry.round()));
+                    signatures.push(SignatureOnG1::from_bytes(entry.signature())?);
+                    validated.push(entry);
+                }
+            }
+
+            pk.verify_batch(
+                messages.iter().map(AsRef::as_ref).collect_vec().as_slice(),
+                signatures.iter().collect_vec().as_slice(),
+            )
+        } else {
+            let mut messages = vec![];
+            let mut signatures = vec![];
+
+            let pk = PublicKeyOnG1::from_bytes(&self.public_key)?;
+            {
+                let cache = self.verified_beacons.read();
+                for curr in entries.iter() {
+                    if prev.round() > 0 && !cache.contains(&curr.round()) {
+                        messages.push(BeaconEntry::message_chained(curr.round(), prev.signature()));
+                        signatures.push(SignatureOnG2::from_bytes(curr.signature())?);
+                        validated.push(curr);
+                    }
+
+                    prev = curr;
+                }
+            }
 
-        // Hash the messages (H(prev sig | curr_round))
-        let digest = {
-            let mut round_bytes = [0; std::mem::size_of::<u64>()];
-            BigEndian::write_u64(&mut round_bytes, curr.round());
-            let mut hasher = sha2::Sha256::default();
-            hasher.update(prev.data());
-            hasher.update(round_bytes);
-            hasher.finalize()
+            verify_messages_chained(
+                &pk,
+                messages.iter().map(AsRef::as_ref).collect_vec().as_slice(),
+                &signatures,
+            )
         };
-        // Signature
-        let sig = Signature::from_bytes(curr.data())?;
-        let sig_match = bls_signatures::verify_messages(&sig, &[&digest], &[self.pub_key.key()?]);
-
-        // Cache the result
-        let contains_curr = self.local_cache.read().contains_key(&curr.round());
-        if sig_match && !contains_curr {
-            self.local_cache.write().insert(curr.round(), curr.clone());
+
+        if is_valid && !validated.is_empty() {
+            let mut cache = self.verified_beacons.write();
+            for entry in validated {
```

Comment:
wibni:
```suggestion
            assert!(cache.cap().get() < validated.len());
            for entry in validated {
```

pathological behaviour for a cache